src/Lean/Meta/ForEachExpr.lean

```diff
@@ -4,47 +4,124 @@ Released under Apache 2.0 license as described in the file LICENSE.
+
+In function applications, each argument is visited, so in `f a b`, all of `f`, `a` and `b` will be visited.
+
+Does not instantiate bound variables if the subexpression is below a binder.
+For that, use `Lean.Meta.visitChildren`.-/
+def Expr.visitChildren [Applicative m] (visit : Expr → m Unit) : Expr → m Unit :=
+  forgetPos Expr.visitChildrenWithPos visit
+
+namespace Meta
+
+variable [Monad m] [MonadLiftT MetaM m] [MonadControlT MetaM m]
+open Lean.SubExpr
+
+/-- Given an expression `e = fun (x₁ : α₁) .. (xₙ : αₙ) => b`, runs `f` on each `αᵢ` and `b`, including a Pos argument. -/
+def visitLambdaWithPos (f : Pos → Expr → m Unit) (p : Pos) (e : Expr) : m Unit := visit #[] p e
+  where visit (fvars : Array Expr) (p : Pos) : Expr → m Unit
     | Expr.lam n d b c => do
       let d := d.instantiateRev fvars
-      f d
+      f p.pushBindingDomain d
       withLocalDecl n c d fun x =>
-        visit (fvars.push x) b
+        visit (fvars.push x) p.pushBindingBody b
     | e => do
-      f <| e.instantiateRev fvars
+      f p <| e.instantiateRev fvars
 
-/-- Given an expression `e =  (x₁ : α₁) → .. (xₙ : αₙ) → b`, runs `f` on each `αᵢ` and `b`. -/
-def visitForall (f : Expr → m Unit) (e : Expr) : m Unit := visit #[] e
-  where visit (fvars : Array Expr) : Expr → m Unit
+
+/-- Given an expression `e =  (x₁ : α₁) → .. (xₙ : αₙ) → b`, runs `f` on each `αᵢ` and `b`, including a Pos argument. -/
+def visitForallWithPos (f : Pos → Expr → m Unit) (p : Pos) (e : Expr) : m Unit := visit #[] p e
+  where visit (fvars : Array Expr) (p : Pos) : Expr → m Unit
     | Expr.forallE n d b c => do
       let d := d.instantiateRev fvars
-      f d
+      f p.pushBindingDomain d
       withLocalDecl n c d fun x =>
-        visit (fvars.push x) b
+        visit (fvars.push x) p.pushBindingBody b
     | e => do
-      f <| e.instantiateRev fvars
+      f p <| e.instantiateRev fvars
 
-/-- Given a sequence of let binders `let (x₁ : α₁ := v₁) ... in b`, runs `f` on each `αᵢ`, `vᵢ` and `b`. -/
-def visitLet (f : Expr → m Unit) (e : Expr) : m Unit := visit #[] e
-  where visit (fvars : Array Expr) : Expr → m Unit
+/-- Given a sequence of let binders `let (x₁ : α₁ := v₁) ... in b`, runs `f` on each `αᵢ`, `vᵢ` and `b`, including a Pos argument. -/
+def visitLetWithPos (f : Pos → Expr → m Unit) (p : Pos) (e : Expr) : m Unit := visit #[] p e
+  where visit (fvars : Array Expr) (p : Pos) : Expr → m Unit
     | Expr.letE n d v b _ => do
       let d := d.instantiateRev fvars
       let v := v.instantiateRev fvars
-      f d
-      f v
+      f p.pushLetVarType d
+      f p.pushLetValue v
       withLetDecl n d v fun x =>
-        visit (fvars.push x) b
+        visit (fvars.push x) p.pushLetBody b
     | e => do
-      f <| e.instantiateRev fvars
+      f p <| e.instantiateRev fvars
+
+/-- Given an expression `e = fun (x₁ : α₁) .. (xₙ : αₙ) => b`, runs `f` on each `αᵢ` and `b`. -/
+def visitLambda (f : Expr → m Unit) (e : Expr) := forgetPos visitLambdaWithPos f e
```

Comment:
I am concerned about the performance implications of changes like this one. 
It feels like `forEachExpr'` is getting more abstract and slower.
I am not sure the current code generator is eliminating all the overhead introduced here. I would have to inspect the generated code.
BTW, I think the transition to `visitLambda (f : Expr -> m Unit) ...` had already introduced overhead, and this one is adding even more overhead.

Reply:
I see, I'll revert the changes for now.